### PR TITLE
Cassandra 19975

### PR DESCRIPTION
**Pull request by dcapwell**

### CODE REVIEW

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 1]
```diff
@@ -223,14 +241,7 @@ private enum RemoveType
 
     private Command<State<S>, Void, ?> removeNodeAssassinate(RandomSource rs, State<S> state)
     {
-        //TODO (correctness): assassinate CMS member isn't allowed
-        IntHashSet up = asSet(state.topologyHistory.up());
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:227]: this was fixed by Marcus on trunk after this logic was written, so cleaning it up

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 2]
```diff
@@ -336,12 +353,15 @@ private EnumSet<TopologyChange> possibleTopologyChanges(State<S> state)
         // so up is enough to know the topology size
         int up = state.topologyHistory.up().length;
         int down = state.topologyHistory.down().length;
+        int[] upAndSafe = state.upAndSafe();
         int total = up + down;
         if (total < state.topologyHistory.maxNodes)
             possibleTopologyChanges.add(TopologyChange.AddNode);
-        if (up > state.topologyHistory.quorum())
+        if (upAndSafe.length > 0)
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:362]: there exists at least 1 node that, if lost, would not break quorum...

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 3]
```diff
@@ -336,12 +353,15 @@ private EnumSet<TopologyChange> possibleTopologyChanges(State<S> state)
...
         {
-            if (up > TARGET_RF)
+            // can't remove the node if all nodes are CMS nodes
+            var nonCMS = Sets.difference(asSet(upAndSafe), asSet(state.cmsGroup));
+            if (!nonCMS.isEmpty())
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: if the only nodes that exist are CMS nodes, avoid remove as its blocked...

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 4]
```diff
@@ -464,19 +488,46 @@ public String ipAddress(int nodeNum)
...
+                            return next;
+                        commandsTransformers.remove(self);
+                        int[] up = state.topologyHistory.up();
+                        List<Command<State<S>, Void, ?>> commands = new ArrayList<>();
+                        commands.add(fixDistributedSchemas);
+                        for (String ks : Arrays.asList("system_auth", "system_traces"))
+                        {
+                            int coordinator = rs.pickInt(up);
+                            commands.add(repairCommand(coordinator, ks));
+                        }
```
Comment (dcapwell) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:527]: if there is data you *need* to repair to make things safe... so added the repairs for that reason

File: test/harry/main/org/apache/cassandra/harry/sut/TokenPlacementModel.java [hunk 5]
```diff
@@ -150,12 +150,12 @@ public interface CompareTo<V>
         int compareTo(V v);
     }
 
-    public static void addIfUnique(List<Replica> replicas, Set<Integer> names, Replica replica)
+    private static void addIfUnique(List<Replica> replicas, Set<String> names, Replica replica)
     {
-        if (names.contains(replica.node().idx()))
+        if (names.contains(replica.node().id()))
```
Comment (dcapwell) [inline test/harry/main/org/apache/cassandra/harry/sut/TokenPlacementModel.java:166]: in this call path `idx() == 0` for every replica, as they don't *share* the same lookup...  personally a few methods are unsafe with this class, but didn't want to refactor and just needed to make safe, so use `id` which is unique regardless of how you create the type

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java (hunk 3 above)
Comment (ifesdjeen) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: Nit: could probably fold `var` into `if`.

Comment (ifesdjeen) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java]: Also, _technically_ we could resize CMS, but we don't have to do this now.

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 6]
```diff
@@ -400,16 +422,20 @@ protected static class State<S extends SchemaSpec> implements AutoCloseable
         final TopologyHistory topologyHistory;
         final Cluster cluster;
         final S schemaSpec;
+        final List<BiFunction<State<S>, Gen<Command<State<S>, Void, ?>>, Gen<Command<State<S>, Void, ?>>>> commandsTransformers = new ArrayList<>();
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:429]: Nit: should we maybe alias this to something? Like an interface, maybe Command Generator and a named function.

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java [hunk 7]
```diff
@@ -464,19 +490,46 @@ public String ipAddress(int nodeNum)
...
-                        preActions.remove(this);
-                    }
+                    });
+                    SimpleCommand<State<S>> fixDistributedSchemas = new SimpleCommand<>("Set system distributed keyspaces to RF=" + TARGET_RF, ignore ->
+                            fixDistributedSchemas(cluster));
+                    SimpleCommand<State<S>> fixTestKeyspace = new SimpleCommand<>("Set " + KEYSPACE + " keyspace to RF=" + TARGET_RF, s -> {
+                        cluster.schemaChange("ALTER KEYSPACE " + KEYSPACE + " WITH replication = {'class': 'SimpleStrategy', 'replication_factor': " + TARGET_RF + "}");
+                        rf = new TokenPlacementModel.SimpleReplicationFactor(TARGET_RF);
+                    });
+                    var self = this;
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:514]: I _think_ you can do `this` without this ref by using `<name_of_enclosing_class>.this`

**Review by ifesdjeen — APPROVED** [review]

File: test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java (hunk 7 above)
Comment (dcapwell) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:514]: Self is smaller ;)

Comment (dcapwell) [reply test/distributed/org/apache/cassandra/fuzz/topology/TopologyMixupTestBase.java:514]: also the enclosing class is an anonymous one `new BiFunction`...

**Pull request merged**